src/radical/pilot/agent/scheduler/base.py

```diff
@@ -477,133 +476,316 @@ def _schedule_units(self, units):
+        incoming tasks to schedule, and a queue of slots freed by finishing
+        tasks.
         '''
 
-        # needs to be locked as we try to acquire slots here, but slots are
-        # freed in a different thread.  But we keep the lock duration short...
-        with self._slot_lock:
+        # The loop alternates between
+        #
+        #   - scheduling tasks from a waitlist;
+        #   - pulling new tasks to schedule; and
+        #   - pulling for free slots to use.
+        #
+        # resources = True  # fresh start
+        #
+        # while True:
+        #
+        #   if resources:  # otherwise: why bother?
+        #     if waitlist:
+        #       sort(waitlist)  # largest tasks first
+        #       for task in sorted(waitlist):
+        #         if task >= max_task:
+        #           prof schedule_skip
+        #         else:
+        #           if try_schedule:
+        #             advance
+        #             continue
+        #           max_task = max(max_task, size(task))
+        #         break  # larger tasks won't work
+        #
+        #   if any_resources:  # otherwise: why bother
+        #     for task in queue_units.get():
+        #         if task <= max_task:
+        #           if try_schedule:
+        #             advance
+        #             continue
+        #           waitlist.append(task)
+        #           max_task = max(max_task, size(task))
+        #         continue  # next task mght be smaller
+        #
+        #   resources = False  # nothing in waitlist fits
+        #   for slot in queue_slots.get():
+        #     free_slot(slot)
+        #     resources = True  # maybe we can place a task now
+
 
-            self._prof.prof('schedule_try', uid=unit['uid'])
-            unit['slots'] = self._allocate_slot(unit['description'])
+        # register unit output channels
+        self.register_output(rps.AGENT_EXECUTING_PENDING,
+                             rpc.AGENT_EXECUTING_QUEUE)
 
-        # the lock is freed here
-        if not unit['slots']:
+        resources = True  # fresh start
+        while not self._proc_term.is_set():
 
-            # signal the unit remains unhandled (Fales signals that failure)
-            self._prof.prof('schedule_fail', uid=unit['uid'])
-            return False
+            active = ''  # see if we do anything in this iteration
 
-        # translate gpu maps into `CUDA_VISIBLE_DEVICES` env
-        self._handle_cuda(unit)
+            # if we have new resources, try to place waiting tasks
+            if resources:
+                a, r = self._schedule_waitlist()
+                if not r: resources  = False
+                if     a: active    += 'w'
 
-        # got an allocation, we can go off and launch the process
-        self._prof.prof('schedule_ok', uid=unit['uid'])
+            # always try to schedule newly incoming tasks
+            a, r = self._schedule_incoming()
```

Comment:
Is this required when `resources is False`?

Reply:
yes - new tasks might be smaller than previous ones.  So if previous tasks could not be placed and we did not get new resources, we might still be able to place smaller incoming ones.
Thanks for checking!